indexer/src/main.rs

```diff
@@ -48,60 +102,43 @@ async fn main() -> anyhow::Result<()> {
-    let exec_plan = Arc::new(ExecutionPlan::load_from_file(&args.plan));
-
-    tracing::info!("{}", "Connecting to database...");
-    let conn = Database::connect(&postgres_url).await?;
+    let Cli { plan, config_path } = Cli::parse();
+
+    tracing::info!("Execution plan {}", plan);
+    let exec_plan = Arc::new(ExecutionPlan::load_from_file(&plan));
+
+    tracing::info!("Config file {:?}", config_path);
+    let file = File::open(&config_path).with_context(|| {
+        format!(
+            "Cannot read config file {path}",
+            path = config_path.display()
+        )
+    })?;
+    let config: Config = serde_yaml::from_reader(file).with_context(|| {
+        format!(
+            "Cannot read config file {path}",
+            path = config_path.display()
+        )
+    })?;
+
+    let mut sink: CardanoSink = match config.sink {
+        SinkConfig::Cardano { .. } => CardanoSink::new(config.sink, exec_plan).await?,
+        _ => todo!("not supported yet"),
+    };
 
-    tracing::info!("{}", "Getting the latest block synced from DB");
+    let start_from = sink.start_from(config.start_block).await?;
 
-    // For rollbacks
-    let points = &match &args.start_block {
-        None => setup::get_latest_points(&conn).await?,
-        Some(block) => setup::get_specific_point(&conn, block).await?,
-    };
-    let intersect = match points {
-        points if points.is_empty() => {
-            // insert genesis then fetch points again
-            genesis::process_genesis(&conn, &network, exec_plan.clone()).await?;
-            // we need a special intersection type when bootstrapping from genesis
-            IntersectArg::Origin
-        }
-        points => {
-            let last_point = points.last().unwrap();
-            tracing::info!(
-                "Starting sync at block #{} ({})",
-                last_point.0,
-                last_point.1
-            );
-            // if last block sync'd was at slot 0,
-            // that means it was the genesis block so we start from origin
-            match last_point.0 {
-                0 => IntersectArg::Origin,
-                _ => IntersectArg::Fallbacks(points.clone()),
-            }
+    let mut engine = match &config.source {
+        SourceConfig::Oura { .. } => {
+            let source = OuraSource::new(config.source, start_from.clone())?;
+            engine::FetchEngine::new(source, sink)
         }
+        _ => todo!("not supported yet"),
     };
 
-    let (handles, input) = setup::oura_bootstrap(intersect, &network, socket)?;
-
-    let sink_setup = postgres_sink::Config { conn: &conn };
-
-    let initial_point = args.start_block.as_ref().map(|_| points.first().unwrap());
-    sink_setup.start(input, exec_plan, initial_point).await?;
-
-    for handle in handles {
-        handle.join().map_err(|_| anyhow!(""))?;
-    }
+    engine
+        .fetch_and_process(start_from.first().unwrap().clone())
+        .await?;
 
-    Ok(())
+    engine.stop().await
```

Comment:
I think this doesn't make a lot of sense. The loop in `fetch_and_process` doesn't exit unless there is an error (would be nice to have the `!` type for that), and if there is an error then the `stop` function will not run.